### PR TITLE
authorization using cancancan gem and role type homepage

### PR DESCRIPTION
**Pull request by lra-wtag**
[Create Candidate Homepage](https://welltravel.atlassian.net/browse/RMS-5)

### CODE REVIEW

File: spec/models/ability_spec.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    end
+    context 'is able to' do
+      let(:candidate) { FactoryBot.create(:candidate, user_id: user.id) }
+      it { is_expected.to be_able_to(:show, candidate) }
+      it { is_expected.to be_able_to(:create, candidate ) }
+      it { is_expected.to be_able_to(:update, candidate ) }
+    end
+
+    context 'is not able to' do
+      let(:candidate) { FactoryBot.create(:candidate, user_id: user.id) }
```
Comment (fnu-wtag) [inline spec/models/ability_spec.rb]: Please remove repeated candidate create. and try to rearrange all the test for a single role.

File: db/migrate/20181115043125_add_role_to_users.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class AddRoleToUsers < ActiveRecord::Migration[5.2]
+  def change
+    add_column :users, :role, :integer, null: false, default: 1
```
Comment (fnu-wtag) [inline db/migrate/20181115043125_add_role_to_users.rb]: why `default: 1`? We set presence validation for role already...

File: spec/models/ability_spec.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      it { is_expected.to_not be_able_to(:edit, candidate) }
+      it { is_expected.to_not be_able_to(:create, candidate ) }
+    end
+  end
+
+  describe 'when user is a Candidate' do
+    let(:user) do
+      FactoryBot.create(:user, role: :candidate)
```
Comment (fnu-wtag) [inline spec/models/ability_spec.rb]: repeated candidate create

File: db/migrate/20181115043125_add_role_to_users.rb (hunk 2 above)
Comment (lra-wtag) [reply db/migrate/20181115043125_add_role_to_users.rb]: as scope says by default an applicant will be created therefore i removed the presence true of role and hence if no role is given by default applicant role will be chosen

File: app/controllers/pages_controller.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
+class PagesController < ApplicationController
+  before_action :check_if_profile_filled?
+
+  def index
+    if current_user.admin?
```
Comment (sde-wtag) [inline app/controllers/pages_controller.rb]: Could we do like this?

```
def index
    if current_user.admin?
      redirect_to candidates_path
    elsif current_user.applicant && current_user.candidate
      render :index
    else
      redirect_to new_candidate_path
    end
  end
```

File: app/models/ability.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  include CanCan::Ability
+
+  def initialize(user)
+    user ||= User.new
+
+    if user.admin?
+      can :manage, :all
+      cannot %i(edit create), Candidate
+    end
+    if user.applicant?
```
Comment (sde-wtag) [inline app/models/ability.rb]: Please use `elsif` condition here.

File: app/views/candidates/_experience.html.erb [hunk 6]
```diff
@@ -2,11 +2,11 @@
 <%= f.input :experience, label: t('shared.all.exp') %>
 <%= f.input :long_term_plan, label: t('shared.all.plan'), input_html: {rows: 3} %>
 <% if @candidate.avatar.attached? && @candidate.avatar.blob_id? %>
-    <%= f.label :avatar, label: t('shared.all.file_up') %>
-    <p class="editfile">
-      <%= @candidate.avatar.filename %>
-    </p>
-    <%= f.check_box :delete_avatar %><%= t('shared.all.delete_file') %><br/>
+    <%= f.label :avatar, label: t('shared.all.file_up') %><%= @candidate.avatar.filename %>
```
Comment (sde-wtag) [inline app/views/candidates/_experience.html.erb]: Fix indentation here.

File: spec/controllers/users_controller_spec.rb [hunk 7]
```diff
@@ -56,4 +47,66 @@
...
     end
   end
+
+  describe 'GET #edit' do
+    before do
+      sign_in_as user
+    end
+    context 'user profile incomplete details' do
+      let(:valid_attributes) do
+        attributes_for(:user,
```
Comment (sde-wtag) [inline spec/controllers/users_controller_spec.rb]: Please use `FactoryBot.attributes_for` here and do same thing where you have written this.

File: spec/factories/candidates.rb [hunk 8]
```diff
@@ -1,18 +1,25 @@
...
+    address { 'house 48, road 8, bashundhara' }
     skill { 'c++,c,ruby' }
     experience { 2.5 }
     personal_interest { 'shopping' }
     hobbies { 'travelling' }
     long_term_plan { 'CEO' }
     keywords { 'software' }
     referrals { 'Tanzim' }
-    email {FFaker::Internet.email}
+    user_id
```
Comment (sde-wtag) [inline spec/factories/candidates.rb]: Please use only `user` here.

File: app/controllers/admins_controller.rb [hunk 9]
```diff
@@ -0,0 +1,7 @@
+class AdminsController < ApplicationController
+  authorize_resource class: AdminsController
+
+  def index
+    render :index
```
Comment (mmo-wtag) [inline app/controllers/admins_controller.rb]: rails will by default render it so no need for this method in fact as rails controller already has these default methods for themselves

File: app/controllers/candidates_controller.rb [hunk 10]
```diff
@@ -13,31 +14,29 @@ def new
...
-      redirect_to candidates_path
+      redirect_to pages_path
     else
       flash.now[:error] = @candidate.errors.full_messages.to_sentence
       render :new
     end
   end
 
-  def edit ; end
+  def edit; end
```
Comment (mmo-wtag) [inline app/controllers/candidates_controller.rb]: you can removed this method as it's empty, rails will already know that you have this method

File: app/controllers/users_controller.rb [hunk 11]
```diff
@@ -1,4 +1,7 @@
 class UsersController < Clearance::UsersController
+  skip_before_action :check_if_profile_filled?, only: %i(edit update)
+  before_action :find_user, only: %i[ edit update ]
```
Comment (mmo-wtag) [inline app/controllers/users_controller.rb]: `load_and_authorize_resource` here also?

File: app/models/user.rb [hunk 12]
```diff
@@ -1,12 +1,20 @@
...
   validates :email,
             presence: true,
             uniqueness: true,
-            format: { with: /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i}
+            format: { with: /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i }
   validates :first_name, :last_name, presence: true
-  validates :contact, presence: true, numericality: true
+
+  def profile_filled?
+    self.country?
```
Comment (mmo-wtag) [inline app/models/user.rb]: no need for this `self` i think?

File: app/views/layouts/application.html.erb [hunk 13]
```diff
@@ -1,48 +1,51 @@
...
+  </head>
+  <body>
+    <div id="header">
+      <%= image_tag('welldevlogo.png', size: '250x40') %>
+      <h1><%= t('layouts.application.title') %></h1>
+      <div id="authentication_control">
+        <% if signed_in? %>
+          <div class="sign_out">
+            <%= image_tag('sign_out.png', size: '23x23', alt: '->') %>
+            <%= link_to 'Sign out', sign_out_path, method: :delete %>
```
Comment (mmo-wtag) [inline app/views/layouts/application.html.erb]: `I18n`?

File: app/views/users/_form.html.erb [hunk 14]
```diff
@@ -3,14 +3,14 @@
...
-<%= f.input :address, label: t('shared.all.address') %>
-<%= f.input :dob, label: t('shared.all.dob'), start_year: Date.today.year,
-            end_year: 1960, order: [:day, :month, :year] %>
 <%= f.country_select :country, { priority_countries: %w(BD),
-                                 prompt: 'Select Country' } %>
+                                 prompt: t('shared.all.country') } %>
 <%= f.input :email, placeholder: 'userdomain.com', label: t('shared.all.email') %>
 <%= f.input :password, label: t('shared.all.password') %>
+<%= f.label :role %>
+<%= f.collection_radio_buttons :role, human_attribute_roles,
```
Comment (mmo-wtag) [inline app/views/users/_form.html.erb]: indentation please

**Pull request merged**